jobmaster/dm/dm_jobmaster.go

```diff
@@ -0,0 +1,246 @@
+	dcontext "github.com/hanfei1991/microcosm/pkg/context"
+	dmpkg "github.com/hanfei1991/microcosm/pkg/dm"
+	"github.com/hanfei1991/microcosm/pkg/p2p"
+)
+
+type JobMaster struct {
+	lib.BaseJobMaster
+
+	workerID libModel.WorkerID
+	jobCfg   *config.JobCfg
+	wg       sync.WaitGroup
+	closeCh  chan struct{}
+
+	metadata              *metadata.MetaData
+	workerManager         *WorkerManager
+	taskManager           *TaskManager
+	messageAgent          *MessageAgent
+	messageHandlerManager p2p.MessageHandlerManager
+}
+
+func RegisterWorker() {
+	constructor := func(ctx *dcontext.Context, id libModel.WorkerID, masterID libModel.MasterID, config lib.WorkerConfig) lib.WorkerImpl {
+		return NewDMJobMaster(ctx, id, masterID, config)
+	}
+	factory := registry.NewSimpleWorkerFactory(constructor, &config.JobCfg{})
+	registry.GlobalWorkerRegistry().MustRegisterWorkerType(lib.DMJobMaster, factory)
+}
+
+func NewDMJobMaster(ctx *dcontext.Context, workerID libModel.WorkerID, masterID libModel.MasterID, conf lib.WorkerConfig) *JobMaster {
+	log.L().Info("new dm jobmaster", zap.String("id", workerID))
+	jm := &JobMaster{
+		workerID: workerID,
+		jobCfg:   conf.(*config.JobCfg),
+		closeCh:  make(chan struct{}),
+	}
+
+	// TODO: we should expose the message handler register Func in base master.
+	// nolint:errcheck
+	ctx.Deps().Construct(func(m p2p.MessageHandlerManager) (p2p.MessageHandlerManager, error) {
+		jm.messageHandlerManager = m
+		return m, nil
+	})
+	return jm
+}
+
+func (jm *JobMaster) createComponments() error {
+	taskStatus, workerStatus, workerHandles, err := jm.getInitStatus()
+	if err != nil {
+		return err
+	}
+	jm.metadata = metadata.NewMetaData(jm.ID(), jm.MetaKVClient())
+	jm.messageAgent = NewMessageAgent(workerHandles, jm.ID(), jm.BaseJobMaster)
+	jm.taskManager = NewTaskManager(taskStatus, jm.metadata.JobStore(), jm.messageAgent)
+	jm.workerManager = NewWorkerManager(workerStatus, jm.metadata.JobStore(), jm.messageAgent, &CheckpointAgentImpl{})
+	return nil
+}
+
+func (jm *JobMaster) InitImpl(ctx context.Context) error {
+	log.L().Info("initializing the dm jobmaster", zap.String("id", jm.workerID), zap.String("jobmaster id", jm.JobMasterID()))
+	if err := jm.createComponments(); err != nil {
+		return err
+	}
+	if err := jm.registerMessageHandler(ctx); err != nil {
+		return err
+	}
+	return jm.taskManager.OperateTask(ctx, Create, jm.jobCfg, nil)
+}
+
+func (jm *JobMaster) Tick(ctx context.Context) error {
+	jm.workerManager.Tick(ctx)
+	jm.taskManager.Tick(ctx)
+	return nil
+}
+
+func (jm *JobMaster) OnMasterRecovered(ctx context.Context) error {
+	log.L().Info("recovering the dm jobmaster", zap.String("id :", jm.workerID))
+	return jm.createComponments()
+}
+
+func (jm *JobMaster) OnWorkerDispatched(worker lib.WorkerHandle, result error) error {
+	if result != nil {
+		log.L().Error("failed to create worker", zap.String("worker_id", worker.ID()), zap.Error(result))
+		jm.workerManager.removeWorkerStatusByWorkerID(worker.ID())
```

Comment:
maybe we don;t need this because if CreateWorker failed, there's no OnWorkerOnline and the worker is not recorded?

Reply:
We [record the worker](https://github.com/hanfei1991/microcosm/blob/3ed20d9ee8677fb1d52af5fc3c2c15c00c19c144/jobmaster/dm/worker_manager.go#L249) after createworker，if a worker create failed，we should remove it.

Reply:
seems [record the worker](https://github.com/hanfei1991/microcosm/blob/3ed20d9ee8677fb1d52af5fc3c2c15c00c19c144/jobmaster/dm/worker_manager.go#L249) may race with OnWorkerDispatched? 

If OnWorkerDispatched comes before [record the worker](https://github.com/hanfei1991/microcosm/blob/3ed20d9ee8677fb1d52af5fc3c2c15c00c19c144/jobmaster/dm/worker_manager.go#L249), we'll left an orphan entry

Reply:
I think this requires framework guarantees? cc @liuzix @amyangfei

Reply:
add create time check in [8106d4f](https://github.com/hanfei1991/microcosm/pull/288/commits/8106d4fc19c911badf59b348ad0de3a17afa6cf9)